packages/swingset-liveslots/src/virtualObjectManager.js

```diff
@@ -586,273 +661,282 @@ export function makeVirtualObjectManager(
-      }
+    if (isDurable) {
+      checkAndUpdateFacetiousness(
+        tag,
+        durableKindDescriptor,
+        facetNames,
+        saveDurableKindDescriptor,
+      );
     }
 
-    vrm.registerKind(kindID, reanimate, deleteStoredVO, isDurable);
-    vrm.rememberFacetNames(kindID, facetNames);
-    harden(contextMapTemplate);
-    harden(prototypeTemplate);
-
-    function makeRepresentative(innerSelf, initializing) {
-      innerSelf.repCount === 0 ||
-        Fail`${innerSelf.baseRef} already has a representative`;
-      innerSelf.repCount += 1;
+    harden(stateShape);
+    stateShape === undefined ||
+      passStyleOf(stateShape) === 'copyRecord' ||
+      assert.fail(X`A stateShape must be a copyRecord: ${q(stateShape)}`);
+    assertPattern(stateShape);
 
-      function ensureState() {
-        if (innerSelf.rawState) {
-          cache.refresh(innerSelf);
-        } else {
-          innerSelf = cache.lookup(innerSelf.baseRef, true);
-        }
-      }
+    let isStateProperty = _prop => undefined;
+    /** @type {(value: any, prop: string) => void} */
+    let checkStateProperty = (_value, _prop) => undefined;
+    if (stateShape) {
+      isStateProperty = prop =>
+        hasOwnPropertyOf(stateShape, prop) ||
+        assert.fail(
+          X`State must only have fields described by stateShape: ${q(
+            ownKeys(stateShape),
+          )}`,
+        );
+      checkStateProperty = (value, prop) => {
+        isStateProperty(prop);
+        mustMatch(value, stateShape[prop]);
+      };
+    }
 
+    // The dataCache holds both unserialized and still-serialized
+    // (capdata) contents of the virtual-object state record.
+    // dataCache[baseRef] -> { capdatas, valueMap }
+    // valueCD=capdatas[prop], value=valueMap.get(prop)
+    /** @type { import('./cache.js').Cache<{ capdatas: any, valueMap: Map<string, any> }>} */
+    const dataCache = makeDataCache(syscall, unserialize);
+    allCaches.push(dataCache);
+
+    // Behavior functions will receive a 'state' object that provides
+    // access to their virtualized data, with getters and setters
+    // backed by the vatstore DB. When those functions are invoked and
+    // we miss in contextCache, we'll call makeState() and
+    // makeContext(). The makeState() call might read from the
+    // vatstore DB if we miss in dataCache.
+
+    // We sample dataCache.get() once each time:
+    // * makeState() is called, which happens the first time in each crank that
+    //   a method is invoked (and the prototype does getContext)
+    // * when state.prop is read, invoking the getter
+    // * when state.prop is written, invoking the setter
+    // This will cause a syscall.vatstoreGet only once per crank.
+
+    function makeState(baseRef) {
       const state = {};
-      if (!initializing) {
-        ensureState();
-      }
-      for (const prop of Object.getOwnPropertyNames(innerSelf.rawState)) {
+      for (const prop of Object.getOwnPropertyNames(
```

Comment:
`Object.getOwnPropertyNames` should probably be replaced with `ownKeys` throughout the file.
```suggestion
      for (const prop of ownKeys(
```

Reply:
I tried that, but tsc started complaining that "`symbol | string` cannot be assigned to `string`". I think we were using `getOwnPropertyNames` because it excludes symbols, whereas `ownKeys` provides both. I'll skip this change.